crates/subspace-service/src/sync_from_dsn/fast_sync.rs

```diff
@@ -0,0 +1,581 @@
+            client,
+            import_queue_service,
+            network_service,
+            fast_sync_state,
+            sync_service,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Sync algorithm:
+    /// - 1. download two last segments,
+    /// - 2. add the last block the second last segment (as raw - without checks and execution),
+    /// - 3. download the state for the first block of the last segment,
+    ///     - add the first block of the last segment as raw,
+    ///     - download state for it,
+    /// - 4. add blocks with execution from the last segment,
+    ///     - notify Substrate sync by importing block with using BlockOrigin::NetworkBroadcast,
+    ///     - clear the block gap to prevent Substrate sync to download blocks from the start.
+    ///     - restart the Substrate sync with SyncMode::Full to enable it downloading full blocks.
+    ///  Note: retry for fast sync will degrade reputation if we have already announced blocks.
+    // TODO: fast-sync specification contains a special case for a segment that have the
+    // complete last archived block, this will remove the necessity to download the second last
+    // segment, we need to implement this case when the blockchain will contain at least one such
+    // a segment to verify the solution.
+    pub(crate) async fn sync(&self) -> Result<FastSyncResult<Block>, Error> {
+        debug!("Starting fast sync...");
+
+        // 1. Download the last two segments.
+        let mut reconstructor = Reconstructor::new().map_err(|error| error.to_string())?;
+        let segment_header_downloader = SegmentHeaderDownloader::new(self.node);
+
+        if let Err(error) = self
+            .download_segment_headers(&segment_header_downloader)
+            .await
+        {
+            let error = format!("Failed to download segment headers: {}", error);
+            return Err(Error::Other(error));
+        };
+
+        let Some(last_segment_index) = self.segment_headers_store.max_segment_index() else {
+            return Err(Error::Other("Can't get last segment index.".into()));
+        };
+
+        // Skip the fast sync if we lack the minimum required segment number
+        if last_segment_index <= SegmentIndex::ONE {
+            return Ok(FastSyncResult::skipped());
+        }
+
+        let last_segment_header = self
+            .segment_headers_store
+            .get_segment_header(last_segment_index)
+            .expect("Last segment index exists; qed");
+
+        let second_last_segment_index = last_segment_header.segment_index() - SegmentIndex::ONE;
+
+        let second_last_segment_blocks = download_and_reconstruct_blocks(
+            second_last_segment_index,
+            self.piece_getter,
+            &mut reconstructor,
+        )
+        .await?;
+
+        let blocks_in_second_last_segment = second_last_segment_blocks.len();
+        debug!(
+            "Second last segment blocks downloaded (SegmentId={}): {}-{}",
```

Comment:
There is no `SegmentId` in the protocol, there are just segment indices. I'd use `%second_last_segment_index` instead of this. On a similar note I'd use `{}..={}` instead of `{}-{}` for block numbers since we're in Rust.